episodes/scraping-data.md

```diff
@@ -6,24 +6,581 @@ exercises: 0
+ <a class="ico zip" href="archive/xls/f923_2021.zip" title="2021"><span>ZIP</span></a>,
+ ...
+ <a class="ico zip" href="archive/xls/f906920_2003.zip" title="2003"><span>ZIP</span></a>,
+ <a class="ico zip" href="archive/xls/f906920_2002.zip" title="2002"><span>ZIP</span></a>,
+ <a class="ico zip" href="archive/xls/f906920_2001.zip" title="2001"><span>ZIP</span></a>,
+ <a class="ico zip" href="archive/xls/f906nonutil2000.zip" title="2000"><span>ZIP</span></a>,
+ <a class="ico zip" href="archive/xls/f906nonutil1999.zip" title="1999"><span>ZIP</span></a>,
+ <a class="ico zip" href="archive/xls/f906nonutil1989.zip" title="1989-1998"><span>ZIP</span></a>]
+```
+
+
+We probably want to skip those Form 906 links too.
+
+```python
+eia_923_links = []
+for a in a_with_zip:
+  if "f906" not in a["href"].lower():
+    eia_923_links.append(a)
+
+eia_923_links
+```
+
+:::: challenge
+
+#### Challenge: only get the relevant `a` tags
+
+Let's grab only the `a` tags that are links to the actual data files.
+
+This will look a little bit different from the 923 example, since the links are all to XLS files instead of ZIP files.
+
+:::::::: solution
+
+```python
+import bs4
+import requests
+
+eia_906_url = "https://www.eia.gov/electricity/data/eia923/eia906u.php"
+response = requests.get(eia_906_url)
+soup = bs4.BeautifulSoup(response.text)
+
+a_tags = soup.find_all("a", href=True)
+eia_906_links = []
+for a in a_tags:
+    if ".xls" in a["href"].lower():
+        eia_906_links.append(a)
+```
+::::::::
+::::
+
+OK, now we have our tags, time to use them to download the data! The EIA 923 data all comes in ZIP files - we'll skip over the handling of those for now. Instead we'll use the 906 data that you just got.
+
+```python
+import pandas as pd
+
+eia_923_blobs = []
+for a in eia_923_links:
+  response = requests.get(a["href"])
+  eia_923_blobs.append(response.content)
+```
+
+Oh no! We get an error:
+
+```output
+MissingSchema: Invalid URL 'xls/f923_2024.zip': No scheme supplied. Perhaps you meant https://xls/f923_2024.zip?
+```
+
+Looks like the URL in the `href` is incomplete. It turns out that this is a *relative path* - much like the relative paths you had to deal with when loading data on your computer. The full URL we want is
+`https://www.eia.gov/electricity/data/eia923/xls/f923_2024.zip` - which combines the URL of the page we got the link from (`https://www.eia.gov/electricity/data/eia923/eia906u.php`) with the fragment we got in the `href` (`xls/f923_2024.zip`).
+
+This is a super common thing to have to do, so there's a useful bit of the Python standard library for this: `urllib.parse.urljoin`:
+
+```
+from urllib.parse import urljoin
+
+urljoin(eia_923_url, "xls/f923_2024.zip")
+```
+
+We can also do the string concatenation ourselves with `eia_906_url + "..."`, but there are a surprising amount of details to get wrong here so it's nice to just use the function that works.
+
+Let's add that into our loop:
```

Comment:
If we show this example, then the next challenge is pretty much "copy and paste". But I think it could be good to still show this before asking people to do it... thoughts?